assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +273,131 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
     Ok(constants)
 }
 
-fn parse_hex_params(
-    op: &Token,
-    param_idx: usize,
-    param_str: &str,
-) -> Result<Vec<Felt>, ParsingError> {
+/// Parses a hexadecimal parameter values without delimeter into a u64 values valid for field
+/// element creation.
+fn parse_long_hex_param(op: &Token, param_str: &str) -> Result<Vec<u64>, ParsingError> {
+    let param_idx = 1;
     // handle error cases where the hex string is poorly formed
-    let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
-        if param_str.len() % 2 != 0 {
-            // parameter string is not a valid hex representation
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        true
-    } else {
-        if param_str.len() % HEX_CHUNK_SIZE != 0 {
-            // hex string doesn't contain a valid number of bytes
-            return Err(ParsingError::invalid_param(op, param_idx));
+    if param_str.len() % HEX_CHUNK_SIZE != 0 {
+        // hex string doesn't contain a valid number of bytes
+        return Err(ParsingError::invalid_param(op, param_idx));
+    }
+
+    // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
+    (0..param_str.len())
+        .step_by(HEX_CHUNK_SIZE)
+        .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
+        .collect()
+}
+
+/// Parses a hexadecimal or decimal parameter value into a u64 value valid for field element
+/// creation.
+fn parse_hex_or_decimal_param(op: &Token, param_idx: usize) -> Result<u64, ParsingError> {
+    debug_assert!(op.num_parts() > param_idx, "param index out of bounds");
+    let param_str = op.parts()[param_idx];
+    let param_value = match param_str.strip_prefix("0x") {
+        Some(param_str) => {
+            debug_assert!(param_str.len() <= HEX_CHUNK_SIZE);
+            parse_hex_param(op, param_idx, param_str)
         }
-        false
+        None => parse_decimal_param(op, param_idx),
     };
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+    param_value.map_err(|_| ParsingError::invalid_param(op, param_idx))
+}
+
+/// Parses a hexadecimal parameter value into a u64 value valid for field element creation.
+fn parse_hex_param(op: &Token, param_idx: usize, param_str: &str) -> Result<u64, ParsingError> {
+    debug_assert_eq!(param_str.len() % 2, 0, "hexadecimal string should be even");
+    match u64::from_str_radix(param_str, 16) {
+        Ok(value) => {
+            if value >= Felt::MODULUS {
+                return Err(ParsingError::invalid_param_with_reason(
+                    op,
+                    param_idx,
+                    format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+                ));
+            }
+            Ok(value)
+        }
```

Comment:
Not introduced in this PR, so maybe unrelated.

Shouldn't we wrap over the modulus instead of err? Meaning: `value = value % Felt::MODULUS`. I understand it might create some confusion, but I think its logical that we either wrap the modulus, or ask the user to create a word that will hold the value

Reply:
I'm not sure I have the right to decide how to do it in this case. @bobbinth what do you think?

Reply:
My thinking here was that if somehow a value greater than `MODULUS` is provided, this is likely to be done in error rather than intentionally. So, we should error out so that the mistake can be caught rather than be silently swallowed.